src/compiler/checker.ts

```diff
@@ -6571,9 +6571,17 @@ namespace ts {
 
         function isJSDocOptionalParameter(node: ParameterDeclaration) {
             return isInJavaScriptFile(node) && (
+                // node.type should only be a JSDocOptionalType when node is a parameter of a JSDocFunctionType
                 node.type && node.type.kind === SyntaxKind.JSDocOptionalType
                 || getJSDocParameterTags(node).some(({ isBracketed, typeExpression }) =>
-                    isBracketed || !!typeExpression && typeExpression.type.kind === SyntaxKind.JSDocOptionalType));
+                    isBracketed || !!typeExpression && skipJSDocPrefixTypes(typeExpression.type).kind === SyntaxKind.JSDocOptionalType));
+        }
+
+        function skipJSDocPrefixTypes(type: TypeNode): TypeNode {

```

Comment:
In `?!a=`, maybe this should be parsed as `(?!a)=` and this wouldn't be necessary?

Reply:
I considered that and will look at that next. The reason is that, for `?T=` the parser does
1. parsePostfix
2. parseNonArrayType
3. parseJSDocNullableType
4. parseType
5. ...
6. parsePostfix

parsePostfix at step (1) doesn't parse the trailing `=` in this case like we would want, the one at step (6) does. This might be fixed by changing step (4) to be parseNonArrayType. I'll try that and see what breaks.

Reply:
In fact, parseJSDocNodeWithType(SyntaxKind.JSDocNonNullable) does this already, so it's probably a mistake that parseJSDocNullableType doesn't.

Reply:
This breaks the existing parsing test in a way that I *think* is good (`...` is no longer allowed in the middle of a type), but I want to test more and maybe revise the parsing of `...` too. I'll do that in a separate PR.

Reply:
Hmm, it also changes how tightly `?T` binds -- if it's tighter than `=`, then it's tighter than `[]` too. So `?T[]` is now `(T | null)[]` where before it was `T[] | null`.